### DISCUSSION TITLE
Unsupported operation

### DISCUSSION THREAD

**Discussion by SandeepJagdev** [pschatzmann/arduino-audio-tools#553] [Q&A] [answered]
hello, everytime i am going to add effect on A2DP it shows unsupported operation in serial monitor. my code is
#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"
#include "AudioLibs/AudioA2DP.h"

// Audio Format
const int sample_rate = 44100;
const int channels = 1;
const int bits_per_sample = 16;

int16_t tremoloDuration = 5;
float tremoloDepth = 0;
Tremolo tremolo(tremoloDuration, tremoloDepth, sample_rate);
BluetoothA2DPSink a2dp_sink;
AudioKitStream kit;
// Access I2S as stream
AudioEffects<GeneratorFromStream<effect_t>> effects(kit, channels , 1.0); // apply effects on kit
GeneratedSoundStream<int16_t> in(effects);
StreamCopy copier(kit, in); // copy in to kit
// Write data to AudioKit in callback
void read_data_stream(const uint8_t *data, uint32_t length) {
    in.write(data, length);
}

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);
 effects.addEffect(tremolo);
  tremolo.setDepth(80);
  tremolo.setDuration(1000);
  tremolo.setActive(tremoloDuration>0);
  // setup output
  auto cfg = kit.defaultConfig(RXTX_MODE);
  cfg.sd_active = false;
  kit.begin(cfg);
 // register callback
  a2dp_sink.set_stream_reader(read_data_stream, false);
  a2dp_sink.start("AudioKit");
}
void loop() {
  kit.processActions();
   copier.copy();
}

If this is not correct method for generate audio-effect then how can i apply echo using  stk on kit .

**Comment by pschatzmann** [comment, marked as answer]
I suggest that you follow the advice given in the [Wiki](https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working)

Just a couple of comments:
- The read_data_callback is supposed to provide the requested data, so I would expect to see a read and definitly not a write
- Double check if you need some special logic to handle the number of channels: A2DP needs 2 channels.
